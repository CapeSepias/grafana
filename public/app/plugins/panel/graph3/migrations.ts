import { FieldConfig, FieldConfigSource, NullValueMode, PanelModel, fieldReducers } from '@grafana/data';
import { GraphFieldConfig, LegendDisplayMode } from '@grafana/ui';
import { AxisPlacement, DrawStyle, LineInterpolation } from '@grafana/ui/src/components/uPlot/config';
import { Options } from './types';
import omitBy from 'lodash/omitBy';
import isNil from 'lodash/isNil';
import { isNumber, isString } from 'lodash';

/**
 * This is called when the panel changes from another panel
 */
export const graphPanelChangedHandler = (
  panel: PanelModel<Partial<Options>> | any,
  prevPluginId: string,
  prevOptions: any
) => {
  // Changing from angular/flot panel to react/uPlot
  if (prevPluginId === 'graph' && prevOptions.angular) {
    const { fieldConfig, options } = flotToGraphOptions(prevOptions.angular);
    panel.fieldConfig = fieldConfig; // Mutates the incoming panel
    return options;
  }

  return {};
};

export function flotToGraphOptions(angular: any): { fieldConfig: FieldConfigSource; options: Options } {
  const overrides = angular.fieldConfig?.overrides ?? [];
  const yaxes = angular.yaxes ?? [];
  let y1 = getFieldConfigFromOldAxis(yaxes[0]);
  if (angular.fieldConfig?.defaults) {
    y1 = {
      ...angular.fieldConfig?.defaults,
      ...y1, // Keep the y-axis unit and custom
    };
  }
  //const y2 = getFieldConfigFromOldAxis(yaxes[1]);

  const graph = y1.custom ?? ({} as GraphFieldConfig);
  graph.drawStyle = angular.bars ? DrawStyle.Bars : angular.lines ? DrawStyle.Line : DrawStyle.Points;
  graph.lineWidth = angular.lineWidth;
  graph.pointSize = angular.pointradius;
  graph.fillOpacity = angular.fill;
  graph.spanNulls = angular.nullPointMode === NullValueMode.Null;
  if (angular.steppedLine) {
    graph.lineInterpolation = LineInterpolation.StepAfter;
  }
  y1.custom = omitBy(graph, isNil);
  y1.nullValueMode = angular.nullPointMode as NullValueMode;

  const options: Options = {
    graph: {},
    legend: {
      displayMode: LegendDisplayMode.List,
      placement: 'bottom',
    },
    tooltipOptions: {
      mode: 'single',
    },
  };

  if (angular.legend?.values) {
    const show = getReducersFromLegend(angular.legend?.values);
    console.log('Migrate Legend', show);
  }

  return {
    fieldConfig: {
      defaults: omitBy(y1, isNil),
      overrides,
    },
    options,
  };
}

// {
//   "label": "Y111",
//   "show": true,
//   "logBase": 10,
//   "min": "0",
//   "max": "1000",
//   "format": "areaMI2",
//   "$$hashKey": "object:19",
//   "decimals": 3
// },
function getFieldConfigFromOldAxis(obj: any): FieldConfig<GraphFieldConfig> {
  if (!obj) {
    return {};
  }
  const graph: GraphFieldConfig = {
    axisPlacement: obj.show ? AxisPlacement.Auto : AxisPlacement.Hidden,
  };
  if (obj.label) {
    graph.axisLabel = obj.label;
  }
  return omitBy(
    {
      unit: obj.format,
      decimals: validNumber(obj.decimals),
      min: validNumber(obj.min),
      max: validNumber(obj.max),
      custom: graph,
    },
    isNil
  );
}

function validNumber(val: any): number | undefined {
  if (isNumber(val)) {
    return val;
  }
  if (isString(val)) {
    const num = Number(val);
    if (!isNaN(num)) {
      return num;
    }
  }
  return undefined;
}

function getReducersFromLegend(obj: Record<string, any>): string[] {
  const ids: string[] = [];
  for (const key of Object.keys(obj)) {
    const r = fieldReducers.getIfExists(key);
    if (r) {
      ids.push(r.id);
    }
  }
  return ids;
}
