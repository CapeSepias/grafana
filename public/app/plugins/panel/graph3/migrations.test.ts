import { PanelModel } from '@grafana/data';
import { graphPanelChangedHandler } from './migrations';

const stairscase = {
  aliasColors: {},
  dashLength: 10,
  fill: 5,
  fillGradient: 6,
  legend: {
    avg: true,
    current: true,
    max: true,
    min: true,
    show: true,
    total: true,
    values: true,
    alignAsTable: true,
  },
  lines: true,
  linewidth: 1,
  nullPointMode: 'null',
  options: {
    alertThreshold: true,
  },
  pointradius: 2,
  seriesOverrides: [],
  spaceLength: 10,
  steppedLine: true,
  thresholds: [],
  timeRegions: [],
  title: 'Panel Title',
  tooltip: {
    shared: true,
    sort: 0,
    value_type: 'individual',
  },
  type: 'graph',
  xaxis: {
    buckets: null,
    mode: 'time',
    name: null,
    show: true,
    values: [],
  },
  yaxes: [
    {
      $$hashKey: 'object:42',
      format: 'short',
      label: null,
      logBase: 1,
      max: null,
      min: null,
      show: true,
    },
    {
      $$hashKey: 'object:43',
      format: 'short',
      label: null,
      logBase: 1,
      max: null,
      min: null,
      show: true,
    },
  ],
  yaxis: {
    align: false,
    alignLevel: null,
  },
  timeFrom: null,
  timeShift: null,
  bars: false,
  dashes: false,
  hiddenSeries: false,
  percentage: false,
  points: false,
  stack: false,
  decimals: 1,
  datasource: null,
};

describe('Graph Migrations', () => {
  it('simple bars', () => {
    const old: any = {
      angular: {
        bars: true,
      },
    };
    const panel = {} as PanelModel;
    const options = graphPanelChangedHandler(panel, 'graph', old);
    expect(panel).toMatchInlineSnapshot(`
      Object {
        "fieldConfig": Object {
          "defaults": Object {
            "custom": Object {
              "lineWidth": undefined,
              "mode": "bars",
              "pointSize": undefined,
            },
            "decimals": undefined,
            "nullValueMode": undefined,
          },
          "overrides": Array [],
        },
      }
    `);
    expect(options).toMatchInlineSnapshot(`
      Object {
        "graph": Object {},
        "legend": Object {
          "displayMode": "list",
          "placement": "bottom",
        },
        "tooltipOptions": Object {
          "mode": "single",
        },
      }
    `);
  });

  it('stairscase', () => {
    const old: any = {
      angular: stairscase,
    };
    const panel = {} as PanelModel;
    const options = graphPanelChangedHandler(panel, 'graph', old);
    expect(panel).toMatchInlineSnapshot(`
      Object {
        "fieldConfig": Object {
          "defaults": Object {
            "custom": Object {
              "lineInterpolation": "staircase",
              "lineWidth": undefined,
              "mode": "line",
              "pointSize": 2,
            },
            "decimals": 1,
            "nullValueMode": "null",
          },
          "overrides": Array [],
        },
      }
    `);
    expect(options).toMatchInlineSnapshot(`
      Object {
        "graph": Object {},
        "legend": Object {
          "displayMode": "list",
          "placement": "bottom",
        },
        "tooltipOptions": Object {
          "mode": "single",
        },
      }
    `);
  });
});
