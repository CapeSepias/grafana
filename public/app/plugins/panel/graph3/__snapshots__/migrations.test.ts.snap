// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph Migrations simple bars 1`] = `
Object {
  "fieldConfig": Object {
    "defaults": Object {
      "custom": Object {
        "drawStyle": "bars",
        "spanNulls": false,
      },
    },
    "overrides": Array [],
  },
  "options": Object {
    "graph": Object {},
    "legend": Object {
      "displayMode": "list",
      "placement": "bottom",
    },
    "tooltipOptions": Object {
      "mode": "single",
    },
  },
}
`;

exports[`Graph Migrations stairscase 1`] = `
Object {
  "fieldConfig": Object {
    "defaults": Object {
      "custom": Object {
        "axisPlacement": "hidden",
        "drawStyle": "line",
        "fillOpacity": 5,
        "lineInterpolation": "stepAfter",
        "pointSize": 2,
        "spanNulls": true,
      },
      "displayName": "DISPLAY NAME",
      "nullValueMode": "null",
      "unit": "short",
    },
    "overrides": Array [],
  },
  "options": Object {
    "graph": Object {},
    "legend": Object {
      "displayMode": "list",
      "placement": "bottom",
    },
    "tooltipOptions": Object {
      "mode": "single",
    },
  },
}
`;

exports[`Graph Migrations twoYAxis 1`] = `
Object {
  "fieldConfig": Object {
    "defaults": Object {
      "custom": Object {
        "axisLabel": "Y111",
        "axisPlacement": "auto",
        "drawStyle": "line",
        "fillOpacity": 1,
        "pointSize": 2,
        "spanNulls": true,
      },
      "decimals": 3,
      "max": 1000,
      "min": 0,
      "nullValueMode": "null",
      "unit": "areaMI2",
    },
    "overrides": Array [],
  },
  "options": Object {
    "graph": Object {},
    "legend": Object {
      "displayMode": "list",
      "placement": "bottom",
    },
    "tooltipOptions": Object {
      "mode": "single",
    },
  },
}
`;
